core/tests/test_utils.py

```diff
@@ -394,19 +394,22 @@ def _parse_json_response(self, json_response, expect_errors=False):
         """Convert a JSON server response to an object (such as a dict)."""
         if not expect_errors:
             self.assertEqual(json_response.status_int, 200)
-
         self.assertEqual(
             json_response.content_type, 'application/json')
         self.assertTrue(json_response.body.startswith(feconf.XSSI_PREFIX))
 
         return json.loads(json_response.body[len(feconf.XSSI_PREFIX):])
 
-    def get_json(self, url, params=None, expect_errors=False):
+    def get_json(self, url, params=None, expect_errors=False,
+                 expected_status_int=200):
         """Get a JSON response, transformed to a Python object."""
         json_response = self.testapp.get(
-            url, params, expect_errors=expect_errors)
-        return self._parse_json_response(
-            json_response, expect_errors=expect_errors)
+            url, params, expect_errors=expect_errors,
+            status=expected_status_int)
+        if (not expect_errors or
```

Comment:
Nothing is returned if the condition fails. It just verifies that the expected_status _int is the status code of the response. If the condition is valid, it goes ahead and returns you the obtained json response.

Reply:
I'll change the docstring to indicate this.

Reply:
But shouldn't it raise an exception or something?

Reply:
Actually no. In some get requests which result in error codes (4xx), the response is a html document. In these cases no testing will be done on the response content, however we check that the status code matches the expected status. We shouldn't raise an exception as the test still makes sense right?

Reply:
Oh, that's kind of unexpected.

Does self.testapp.get() error (as in, raise an Exception) when expected_status_int is not the same as the observed status? If so, that's good; if not, we need a check somewhere.

In the case where there's an error and the content type is HTML, I would say we should return the error page. But something weird is going on here. Why wouldn't the response be application/json? After all this is called get_json() which suggests it's meant for JSON handlers in the first place.

Reply:
yes testapp.get() raises errors if status doesn't match. I thought the same about not getting JSON responses from get_json (But when I assumed it would be a JSON response and tried to use it in that way even for error scenarios, I ran into a lot of backend failures. Seems like this is used in a lot of places with HTML responses.)

Reply:
OK, well, for HTML responses can we declare a get_html() instead and use that? Then have this method raise error if return value isn't json, and have the other method raise an error if return value isn't html?

I mean, regardless of success/failure, the type is consistent, right? I.e. we don't have any url choosing to serve html or json based on the status code? If so, that means the caller has an expectation for what format the returned data should be in, and two methods would make sense.

Reply:
This one I'll need to think about. I'll investigate it a bit more and get back.But actually with regard to your second point, the type isn't consistent regardless of success/failure. 200 response resulted in a JSON object while a 401 response resulted in a HTML document for me. This is why I had worked around it in that fashion.

Reply:
Oh, interesting, that actually seems wrong -- thanks for detecting it!

Perhaps we could change the _code_ behavior to be at least consistent in the type of object returned? (I don't really see the point of returning a JSON response when an HTML one is expected, and vice versa.)